File: ABC/abc284/e.py
N, M = map(int, input().split())

edges = []
for i in range(N):
  edges.append([])
for i in range(M):
  u, v = map(int, input().split())
  u -= 1
  v -= 1
  edges[u].append(v)
  edges[v].append(u)

# print(edges)

import sys
sys.setrecursionlimit(10**8)

visited = [False]*N

limit = 10**6
cnt = 0

def dfs(i):
  global cnt
  cnt += 1
  if cnt == limit:
    print(cnt)
    exit()
    # return
  visited[i] = True
  for j in edges[i]:
    if not visited[j]:
      dfs(j)
  visited[i] = False

dfs(0)
print(cnt)

File: ABC/abc284/e.cpp
#include <bits/stdc++.h>
using namespace std;
// #define rep(i,n) for (int i=0; i<(n); i++);
using ll = long long;

int N, M;
int u, v;
int i;
vector<int> G[1000009];
bool visited[1000009];
int cnt = 0;

int dfs(int pos) {
  visited[pos] = true;
  cnt++;
  if (cnt == 1e6) {
    cout << cnt << endl;
    exit(0);
  }
  for(int i=0; i<G[pos].size(); i++){
    int nex = G[pos][i];
    if (visited[nex] == false) dfs(nex);
  }
  visited[pos] = false;
  return cnt;
}

int main(){
  cin >> N >> M;
  for (int i=0; i<M; i++){
    cin >> u >> v;
    u--;
    v--;
    G[u].push_back(v);
    G[v].push_back(u);
  }
  for(int i=0; i<N; i++) visited[i] = false;
  dfs(0);
  cout << cnt << endl;
  return 0;
}